packages/wonder-blocks-timing/src/hooks/use-timeout.ts

```diff
@@ -1,36 +1,73 @@
-) {
-    // We using a ref instead of a callback for `action` to avoid resetting
-    // the interval whenever the `action` changes.
-    const actionRef = useUpdatingRef(action);
+    options: HookOptions = {},
+): ITimeout {
+    const {actionPolicy, clearPolicy, schedulePolicy} = options;
+    const actionProxyRef = useRef<() => unknown>(action);
+    const timeoutRef = useRef<ITimeout | null>(null);
 
-    useEffect(() => {
-        if (active) {
-            const timeoutId = setTimeout(() => {
-                actionRef.current();
-            }, timeoutMs);
-
-            return () => {
-                clearTimeout(timeoutId);
-            };
+    // Since we are passing our proxy function to the timeout instance,
+    // it's check that the action is a function will never fail. So, we have to
+    // do that check ourselves, and we do it here.
+    if (typeof action !== "function") {
+        throw new Error("Action must be a function");
+    }
+
+    // If we're rendered with an updated action, we want to update the ref
+    // so the existing timeout gets the new action, and then reset the
+    // timeout if our action policy calls for it.
+    if (action !== actionProxyRef.current) {
+        actionProxyRef.current = action;
+        if (actionPolicy === ActionPolicy.Reset) {
+            timeoutRef.current?.set();
         }
-        // actionRef isn't actually required, but react-hooks/exhaustive-deps
-        // doesn't recognize it as a ref and thus complains if it isn't in the
-        // deps list.  It isn't a big deal though since the value ofactionRef
-        // never changes (only its contents do).
-    }, [timeoutMs, active, actionRef]);
+    }
+
+    // This effect updates the timeout when the timeoutMs, clearPolicy,
+    // or schedulePolicy changes.
+    useEffect(() => {
+        // Make a new timeout.
+        timeoutRef.current = new Timeout(
+            () => {
+                actionProxyRef.current?.();
+            },
+            timeoutMs,
+            schedulePolicy,
+        );
+
+        // Clear the interval when the effect is cleaned up, if necessary,
+        // making sure to use the clear policy.
+        return () => {
+            timeoutRef.current?.clear(clearPolicy);
+            timeoutRef.current = null;
+        };
+    }, [timeoutMs, clearPolicy, schedulePolicy]);
+
+    // This is the API we expose to the consumer. We expose this rather than
+    // the interval instance itself so that the API we give back is stable
+    // even if the underlying interval instance changes.
+    const externalApi = useMemo(
+        () => ({
+            set: () => {
+                timeoutRef.current?.set();
+            },
+            clear: (policy?: ClearPolicy) => {
+                timeoutRef.current?.clear(policy);
+            },
+            get isSet() {
+                return timeoutRef.current?.isSet ?? false;
+            },
+        }),
+        [],
+    );
+
+    return externalApi;
```

Comment:
As I was looking through this PR it struck me how similar the `useTimeout` and `useInterval` hooks were. We handle things virtually identically between the two implementations. I'm wondering if it'd be worth unifying them into a single underlying implementation and just passing in which `setXyz()` JS function to use to drive the hook (eg. `setTimeout` or `setInterval`). 🤔  Or maybe that just is too much abstraction.

Reply:
Yes, you may be right. We could possibly replace this with a single underlying hook like `useActionScheduler` that takes the type of scheduler somehow. Not sure if it's worth the effort in golfing it down to that though.